1_The_Physical/Level_2_Stateful_Guess_Point/Shared/Student/src/domain/student/entity/Student.ts

```diff
@@ -0,0 +1,133 @@
+import { v4 as uuid } from "uuid";
+
+import { EventBus } from "../../../infra/EventBus";
+import { Result } from "../../../shared/Result";
+import { AggregateRoot } from "../../core/AggregateRoot";
+import { FirstNameUpdated, LastNameUpdated, StudentCreated } from "../events";
+import { Email, FirstName, LastName } from "../value-objects";
+import { FirstNameValidationError } from "../validators/FirstNameValidator";
+import { LastNameValidationError } from "../validators/LastNameValidator";
+import { EmailValidationError } from "../validators/EmailValidator";
+
+interface StudentProps {
+  firstName: string;
+  lastName: string;
+}
+
+interface StudentState {
+  firstName: FirstName;
+  lastName: LastName;
+  email: Email;
+}
+
+interface StudentValidationError {
+  firstName?: FirstNameValidationError;
+  lastName?: LastNameValidationError;
+  email?: EmailValidationError;
+}
+
+export class Student implements AggregateRoot<StudentState> {
+  public readonly id: string;
+  public readonly state: StudentState;
+
+  private constructor(props: StudentState, eventBus?: EventBus) {
+    this.id = uuid();
+    this.state = props;
+    eventBus?.publish(new StudentCreated(this));
```

Comment:
It looks like you don't store events anywhere anymore, but there is a requirement for that:

> When I updateFirstName or updateLastName, the student object saves an event to an events list - this is how we can tell how the object has changed.